components/scream/src/share/io/scream_scorpio_interface.F90

```diff
@@ -1216,61 +1218,84 @@ subroutine grid_write_darray_1d_real(filename, varname, var_data_ptr)
+    else
+      call errorHandle( "eam_grid_write_darray_1d: Error writing variable "//trim(varname)//" unsupported dtype", -999)
+    end if
+  end subroutine grid_write_darray_1d
 !=====================================================================!
-  ! Read output from file based on type (int or real) and dimensionality
-  ! (currently support 1-4 dimensions).
+  ! Read output from file based on type (int or real)
   ! --Note-- that any dimensionality could be read if it is flattened to 1D
   ! before calling a write routine.
   ! Mandatory inputs are:
-  ! filename: the netCDF filename to be read from.
-  ! hbuf:     An array of data that will be used to read the input.  NOTE:
-  !           If PIO MPI ranks > 1, hbuf should be the subset of the global array
-  !           which includes only those degrees of freedom that have been
-  !           assigned to this rank.  See set_dof above.
-  ! varname:  The shortname for the variable being read.
+  ! filename:     the netCDF filename to be read from.
+  ! varname:      The shortname for the variable being read.i
+  ! var_data_ptr: An c_ptr of data that will be used to read the input.  NOTE:
+  !               If PIO MPI ranks > 1, var_data_ptr should be the subset of the global array
+  !               which includes only those degrees of freedom that have been
+  !               assigned to this rank.  See set_dof above.
   !---------------------------------------------------------------------------
   !
-  !  grid_read_darray_1d_real: Read a variable defined on this grid
+  !  grid_read_darray_1d: Read a variable defined on this grid
   !
   !---------------------------------------------------------------------------
-  subroutine grid_read_darray_1d_real(filename, varname, var_data_ptr)
+  subroutine grid_read_darray_1d(filename, varname, var_data_ptr, time_index)
 
     ! Dummy arguments
     character(len=*), intent(in) :: filename       ! PIO filename
     character(len=*), intent(in) :: varname
     type (c_ptr),     intent(in) :: var_data_ptr
+    integer, intent(in)          :: time_index
 
     ! Local variables
     type(pio_atm_file_t),pointer       :: pio_atm_file
     type(hist_var_t), pointer          :: var
     integer                            :: ierr, var_size
     logical                            :: found
-    real(rtype), dimension(:), pointer :: var_data
+    real(rtype), dimension(:), pointer         :: var_data_real
+    integer(kind=c_int), dimension(:), pointer :: var_data_int
 
     call lookup_pio_atm_file(trim(filename),pio_atm_file,found)
     call get_var(pio_atm_file,varname,var)
+
+    ! Set the timesnap we are reading
+    if (time_index .gt. 0) then
+      call PIO_setframe(pio_atm_file%pioFileDesc,var%piovar,int(time_index,kind=pio_offset_kind))
+    else
+      call PIO_setframe(pio_atm_file%pioFileDesc,var%piovar,int(max(1,pio_atm_file%numRecs),kind=pio_offset_kind))
```

Comment:
It seems that `numRecs` is the number of records on file, while `time_index` is the time index since the beginning of the simulation. I am wondering what happens when we limit the number of snapshots on file. Say you put 10 snapshots per file, at the 11th time step, `time_index=11`, but `numRecs=0` on the file. Is it ok to write to file a time index 11, withuot any previous time slice? I guess so, but I want to make sure we are not overlooking some index arithmetics that scorpio might assume under the hood.

Reply:
My guess is that if you ask for a time_index that is outside of the bounds then PIO would return an error.  We can check for that.

I'll note, this is currently only being used for reading, so no writing.  I'm not sure what the policy is for writing when you specify a time index.  If PIO will fill in all missing time_indices with something, or if it will throw an error. 

I'm also not sure if we will ever use a similar syntax for writing, since we can currently already write multiple time snaps per file and we follow the convention of just bumping the numRecs up by one every time we write.

Reply:
You're right this is a read routine. I would then like to add a check before the call to `PIO_setframe`. Even if PIO would error out, I think we can do a better job with the error message, which would make life easier for debugging.

We should check that `0<=time_index<num_time_slices_on_file` (or add 1 on both sides if using netcdf numbering).  Then, if the check fails, we should print a nice msg printing the requested slice, and the valid interval.

Reply:
Okay, this will actually require a bit more work I think.  numRecs is set manually by the interface and only changes when `eam_update_timestep` is called.  Perhaps we should remove it entirely from the read_input call.  And instead query the size of the time dimension directly from PIO and then put in your check.

Reply:
Taking this offline.